pyls/plugins/importmagic_lint.py

```diff
@@ -0,0 +1,264 @@
+            'severity': sevirity_level,
+        }
+
+    Args:
+        document: The document to be linted.
+    Returns:
+        A list of dictionaries.
+    """
+    scope = importmagic.Scope.from_source(document.source)
+    unresolved, unreferenced = scope.find_unresolved_and_unreferenced_symbols()
+
+    diagnostics = []
+
+    # Annoyingly, we only get the text of an unresolved import, so we'll look for it ourselves
+    for unres in unresolved:
+        for line_no, line in enumerate(document.lines):
+            pos = line.find(unres)
+            if pos < 0:
+                continue
+
+            diagnostics.append({
+                'source': SOURCE,
+                'range': {
+                    'start': {'line': line_no, 'character': pos},
+                    'end': {'line': line_no, 'character': pos + len(unres)}
+                },
+                'message': "Unresolved import '%s'" % unres,
+                'severity': lsp.DiagnosticSeverity.Hint,
+            })
+
+    for unref in unreferenced:
+        for line_no, line in enumerate(document.lines):
+            pos = line.find(unref)
+            if pos < 0:
+                continue
+
+            # Find out if the unref is an import or a variable/func
+            imports = _get_imports_list(document.source)
+            if unref in imports:
+                message = "Unreferenced import '%s'" % unref
+            else:
+                message = "Unreferenced variable/function '%s'" % unref
+
+            diagnostics.append({
+                'source': SOURCE,
+                'range': {
+                    'start': {'line': line_no, 'character': pos},
+                    'end': {'line': line_no, 'character': pos + len(unref)}
+                },
+                'message': message,
+                'severity': lsp.DiagnosticSeverity.Warning,
+            })
+
+    return diagnostics
+
+
+@hookimpl
+def pyls_code_actions(config, document, context):
+    """Build a list of actions to be suggested to the user. Each action follow this format:
+        {
+            'title': 'importmagic',
+            'command': command,
+            'arguments':
+                {
+                    'uri': document.uri,
+                    'version': document.version,
+                    'startLine': start_line,
+                    'endLine': end_line,
+                    'newText': text,
+                }
+        }
+    """
+    # Update the style configuration
+    conf = config.plugin_settings('importmagic_lint')
+    min_score = conf.get('minScore', 1)
+    log.debug("Got importmagic settings: %s", conf)
+    importmagic.Imports.set_style(**{_utils.camel_to_underscore(k): v for k, v in conf.items()})
+
+    # Might be slow but is cached once built
+    # TODO (youben): add project path for indexing
+    index = _get_index(sys.path)
+    actions = []
+    diagnostics = context.get('diagnostics', [])
+    for diagnostic in diagnostics:
```

Comment:
I wonder if it's easier to just call `pyls_lint(document)`, then try to find the context diagnostic in the results. That way we don't have to parse the messages, just have to do an equality check.

e.g. `action_diags = [diag for diag in pyls_lint(document) if diag in context.get('diagnostics', [])]`

Reply:
True that the diagnostics list can grow bigger than the local diagnostics and we will just pass most of them. Also we don't need to do the equality check if we call pyls_lint() as all the diagnostics will be processed.